### PR TITLE
allow custom ordering of default panels

### PR DESCRIPTION
**Pull request by VeraZab**

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
To capture our verbal conversation: I think the ideal API here would be to pass in two arrays-of-strings, which would define the desired order, and that would be applied within `PanelMenuWrapper`.

**Comment by VeraZab** [comment]
@dmt0 @nicolaskruchten please review

**Comment by dmt0** [comment]
Code looks good!

**Comment by nicolaskruchten** [comment]
A few test cases would be nice, including some edge cases like missing items in either the desired or actual order, mismatched group/name pairs, that sort of thing. Just to lock down the desired behaviour in a test :)

File: src/components/PanelMenuWrapper.js [hunk 1]
```diff
@@ -61,17 +61,20 @@ class PanelsWithSidebar extends Component {
       const nameOrder = order.map(panel => panel.name).filter(this.getUniqueValues);
 
       orderedChildren.sort((a, b) => {
+        const panelAHasCustomOrder = groupOrder.includes(a.props.group);
```
Comment (dmt0) [inline src/components/PanelMenuWrapper.js]: 👍

File: src/lib/sortMenu.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
+function getUniqueValues(value, index, self) {
+  return self.indexOf(value) === index;
+}
+
+function sortAlphabetically(a, b) {
+  const sortByGroup = a.props.group === b.props.group ? 0 : a.props.group < b.props.group ? -1 : 1;
+  const sortByName = a.props.name === b.props.name ? 0 : a.props.name < b.props.name ? -1 : 1;
+  return sortByGroup || sortByName;
```
Comment (nicolaskruchten) [inline src/lib/sortMenu.js]: clever! took me a sec to think about how that works :)

File: src/lib/__tests__/sortMenu-test.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+describe('sortMenu', () => {
+  it('modifies original array to follow the group, then name order provided', () => {
+    const initialArray = [
+      {props: {group: 'DEV', name: 'Inspector'}},
+      {props: {group: 'DEV', name: 'JSON'}},
+    ];
+    const orderProp = [{group: 'DEV', name: 'JSON'}, {group: 'DEV', name: 'Inspector'}];
+    sortMenu(initialArray, orderProp);
+
+    expect(initialArray[0].props.name).toBe('JSON');
```
Comment (nicolaskruchten) [inline src/lib/__tests__/sortMenu-test.js]: I would express these as `expect(initialArray).toBe( ... the full array ... )` ... More readable and more complete IMO :)

File: src/lib/__tests__/sortMenu-test.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
+import sortMenu from '../sortMenu';
+
+describe('sortMenu', () => {
+  it('modifies original array to follow the group, then name order provided', () => {
+    const initialArray = [
+      {props: {group: 'DEV', name: 'Inspector'}},
```
Comment (nicolaskruchten) [inline src/lib/__tests__/sortMenu-test.js]: I would consider making a helper here that takes in threes arrays of strings of form `GROUP/NAME` and does a split and does the comparison, just to have some very compact test cases below

File: src/lib/sortMenu.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  // eslint-disable-next-line
+  order = order.filter(desiredPanel =>
+    panels.some(
+      actualPanel =>
+        actualPanel.props.name === desiredPanel.name &&
+        actualPanel.props.group === desiredPanel.group
+    )
+  );
+
+  const desiredGroupOrder = order.map(panel => panel.group).filter(getUniqueValues);
```
Comment (nicolaskruchten) [inline src/lib/sortMenu.js]: arguably this is unnecessary right? It won't change the order :)

File: src/DefaultEditor.js [hunk 6]
```diff
@@ -95,6 +95,7 @@ class DefaultEditor extends Component {
 DefaultEditor.propTypes = {
   children: PropTypes.node,
   logoSrc: PropTypes.string,
+  order: PropTypes.array,
```
Comment (nicolaskruchten) [inline src/DefaultEditor.js]: maybe this should be called `menuPanelOrder` or something, for clarity

**Comment by nicolaskruchten** [comment]
💃 ... all comments are just notes, not requests to change
